keps/sig-node/4816-dra-prioritized-list/README.md

```diff
@@ -0,0 +1,1124 @@
+  name: device-consumer-claim
+spec:
+  devices:
+    requests:
+    - name: nic
+      deviceClassName: rdma-nic
+    - name: gpu
+      firstAvailableOf:
+      - name: big-gpu
+        deviceClassName: big-gpu
+      - name: mid-gpu
+        deviceClassName: mid-gpu
+      - name: small-gpu
+        deviceClassName: small-gpu
+        count: 2
+    constraints:
+    - requests: ["nic", gpu"]
+      matchAttribute:
+      - dra.k8s.io/pcieRoot
+    config:
+    - requests: ["small-gpu"]
+      opaque:
+        driver: gpu.acme.example.com
+        parameters:
+          apiVersion: gpu.acme.example.com/v1
+          kind: GPUConfig
+          mode: multipleGPUs
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  resourceClaims:
+  - name: "gpu-and-nic"
+    resourceClaimName: device-consumer-claim
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+      claims:
+      - name: "gpu-and-nic"
+        request: "gpu" # the 'nic' request is pod-level, no need to attach to container
+```
+
+There are a few things to note here. First, the "nic" request is listed with a
+`deviceClassName`, because it has no alternative request types. The "gpu"
+request could be met by several different types of GPU, in the listed order of
+preference. Each of those is a separate `DeviceRequest`, with both a
+`deviceClassName` and also its own name. The fact that these subrequests also
+have their own names allows us to apply constraints or configuration to
+specific, individual subrequests, in the event that it is the chosen
+alternative. In this example, the "small-gpu" choice requires a configuration
+option that the other two choices do not need. Thus, if the resolution of the
+"gpu" request is made using the "small-gpu" subrequest, then that configuration
+will be attached to the allocation. Otherwise, it will not.
+
+Similarly, for `Constraints`, the list of requests can include the main request
+name ("gpu" in this case), in which case the constraint applies regardless of
+which alternative is chosen. Or, it can include the subrequest name, in which
+case that constraint only applies if that particular subrequest is chosen.
+
+In the PodSpec, however, the subrequest names are not valid. Only the main
+request name may be used.
+
+### Scheduler Implementation
+
+Currently, the scheduler loops through each entry in `DeviceClaim.Requests` and
+tries to satisfy each one. This would work essentially the same, except that
+today, it [throws an
+error](https://github.com/kubernetes/kubernetes/blob/03f134461462f86239067ec20ec17a0ba892db52/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L164)
+when it encounters a claim with a missing `DeviceClassName`. Instead, here we
+would check for entries in `FirstAvailableOf`, and add an additional loop,
```

Comment:
What if there is one ResourceSlice (associated to a node) that only has the 2nd device of `firstAvailableOf` available, and another node that has the 1st device of `firstAvailableOf` available. Will the scheduler be instructed to choose the second node? If so, how?

Reply:
That would be via scoring, which we do not currently implement for DRA.

Reply:
Made scoring a beta criteria; for alpha, it would not prefer one node over the other. Maybe we'll try to do that in alpha but for now it's a beta criteria.